src/gateway/environments/jsdom-sixteen/jsdom-sixteen-resource-loader.js

```diff
@@ -0,0 +1,144 @@
+     */
+    constructor(
+        renderAPI: RenderAPI,
+        requestOptions?: RequestOptions = DefaultRequestOptions,
+    ) {
+        // Patch before super to make sure promises get an abort.
+        applyAbortablePromisesPatch();
+
+        super();
+
+        if (renderAPI == null) {
+            throw new Error("Must provide render API.");
+        }
+
+        this._active = true;
+        this._renderAPI = renderAPI;
+        this._requestOptions = requestOptions;
+    }
+
+    get isActive(): boolean {
+        return this._active;
+    }
+
+    close(): void {
+        this._active = false;
+        abortInFlightRequests();
+    }
+
+    fetch(url: string, options?: FetchOptions): ?Promise<Buffer> {
+        const logger = this._renderAPI.logger;
+        const isInlineData = url.startsWith("data:");
+        const readableURLForLogging = isInlineData ? "inline data" : url;
+        if (!this._active) {
+            /**
+             * If we get here, then something is trying to fetch when our
+             * environment has closed us down. This could be in the reject
+             * or resolve of a promise, for example.
+             *
+             * If it's inlinedata, it really doesn't matter, so let's log it
+             * only if it's for a file.
+             */
+            if (!isInlineData) {
+                logger.warn(
+                    `File fetch attempted after resource loader close: ${readableURLForLogging}`,
+                );
+            }
+
+            /**
+             * Though we intentionally don't want to load this file, we can't
+             * just return null per the spec as this can break promise
+             * resolutions that are relying on this file. Instead, we resolve
+             * as an empty string so things can tidy up properly.
+             */
+            return JSDOMSixteenResourceLoader.EMPTY_RESPONSE;
+        }
+
+        /**
+         * We must still be active.
+         * If this request is not a JavaScript file, we are going to return an
+         * empty response as we don't care about non-JS resources.
+         */
+        const JSFileRegex = /^.*\.js(?:\?.*)?/g;
+        if (!JSFileRegex.test(url)) {
+            logger.silly(`EMPTY: ${readableURLForLogging}`);
+
+            /**
+             * Though we intentionally don't want to load this file, we can't
+             * just return null per the spec as this can break promise
+             * resolutions that are relying on this file. Instead, we resolve
+             * as an empty string so things can tidy up properly.
+             */
+            return JSDOMSixteenResourceLoader.EMPTY_RESPONSE;
+        }
+
+        /**
+         * This must be a JavaScript file request. Let's make a request for the
+         * file and then handle it coming back.
+         */
+        const abortableFetch = request(logger, url, {
+            ...this._requestOptions,
```

Comment:
Because the request options configure caching and such, we can guarantee that we share superagent things with any other request calls that use the same options.